File: algorithm/98.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <ctype.h>
#include <math.h>
struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

int dfs(struct TreeNode *root, long long low, long long high)
{
    if (root == NULL)
    {
        return 1;
    }
    if (root->val <= low || root->val >= high)
    {
        return 0;
    }
    return dfs(root->left, low, root->val) && dfs(root->right, root->val, high);
}

int isValidBST(struct TreeNode *root)
{
    return dfs(root, LONG_MIN, LONG_MAX);
}

File: go/goalgor/algor2/98.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

const INT64_MAX = int64(^uint64(0) >> 1)
const INT64_MIN = ^INT64_MAX

func check(root *TreeNode, low, high int64) bool {
	if root == nil {
		return true
	}
	if int64(root.Val) <= low || int64(root.Val) >= high {
		return false
	}
	return check(root.Left, low, int64(root.Val)) && check(root.Right, int64(root.Val), high)
}

func isValidBST(root *TreeNode) bool {
	return check(root, INT64_MIN, INT64_MAX)
}
